### PR TITLE
Added KphpNotSupportFunctionsInspection inspection

### PR DESCRIPTION
**Pull request by Danil42Russia**
# Info
Since KPHP does not support some functions, it would be nice to report this to the IDE
This place is in the KPHP code: [link](https://github.com/VKCOM/kphp/blob/2ba512e9a563a1d2fa3ee39c1002454f5db164fa/compiler/pipes/deduce-implicit-types-and-casts.cpp#L546-L551)

## Example
<details>
<summary>Images</summary>

image

image
</details>

### CODE REVIEW

**Comment by i582** [comment]
Interesting, you can also add a quick fix.

**Comment by Danil42Russia** [comment]
> Interesting, you can also add a quick fix.

#### 1. Select function
image

#### 2. Show Context Action 
image

#### 3. Selecting a replacement function
image

#### 4. Replace function
image

**Comment by i582** [comment]
@Danil42Russia, cool, I like it!
Let's wait what Sasha says about this.

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.openapi.project.Project
+import com.jetbrains.php.lang.lexer.PhpTokenTypes
+import com.jetbrains.php.lang.psi.PhpPsiElementFactory
+import com.jetbrains.php.lang.psi.elements.FunctionReference
+
+class ReplaceToKphpFunctionsQuickFix(private val maybeFunctionName: String) : LocalQuickFix {
+    override fun getFamilyName(): String {
+        return "Replace with $maybeFunctionName()"
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:12]: I think it should be replaced with expression body, as it is used in the rest of the project

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val unsupportedFunction = descriptor.psiElement
+        if (unsupportedFunction !is FunctionReference) {
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:17]: Here you use curly braces for `if` with a single return body, but above you don't. 

Judging by the rest of the project, you usually don't need to use curly braces at all for such `if`s.

File: build.gradle [hunk 3]
```diff
@@ -40,7 +40,7 @@ intellij {
     ]
 }
 runIde {
-    ideDir.set(file("/Users/alexandr.kirsanov/Library/Application Support/JetBrains/Toolbox/apps/PhpStorm/ch-0/213.5744.279/PhpStorm.app/Contents"))
+    // ideDir.set(file("/Users/alexandr.kirsanov/Library/Application Support/JetBrains/Toolbox/apps/PhpStorm/ch-0/213.5744.279/PhpStorm.app/Contents"))
```
Comment (i582) [inline build.gradle:43]: I don't think it should be part of this PR.

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package com.vk.kphpstorm.inspections.quickfixes
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.openapi.project.Project
+import com.jetbrains.php.lang.lexer.PhpTokenTypes
+import com.jetbrains.php.lang.psi.PhpPsiElementFactory
+import com.jetbrains.php.lang.psi.elements.FunctionReference
+
+class ReplaceToKphpFunctionsQuickFix(private val maybeFunctionName: String) : LocalQuickFix {
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:10]: I don't think this is the best field name here.

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val unsupportedFunction = descriptor.psiElement
+        if (unsupportedFunction !is FunctionReference) {
+            return
+        }
+
+        val methodNameNode = unsupportedFunction.nameNode ?: return
+
+        val newFunction = PhpPsiElementFactory.createFromText(
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:23]: Maybe it's better to name the variable `newFunctionNameNode`?

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        return "Replace with $maybeFunctionName()"
+    }
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val unsupportedFunction = descriptor.psiElement
+        if (unsupportedFunction !is FunctionReference) {
+            return
+        }
+
+        val methodNameNode = unsupportedFunction.nameNode ?: return
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:21]: Why a method? We are replacing simple functions, not methods.

**Review by i582 — CHANGES_REQUESTED** [review]

**Comment by i582** [comment]
It seems strange to me that the suggestion to replace with `array_first_key` comes first, since `reset` returns the value of the first element, not the key.

Снимок экрана 2022-02-20 в 16 27 29

**Comment by i582** [comment]
```php
<?php # functions.php

namespace MyFunctions;

function reset() {}
```

```php
<?php # script.php

$a = [];

echo MyFunctions\reset();
```

Снимок экрана 2022-02-20 в 16 31 55

I think it's better to compare function names with fqn to avoid this.

**Comment by Danil42Russia** [comment]
> It seems strange to me that the suggestion to replace with `array_first_key` comes first, since `reset` returns the value of the first element, not the key.

#### Plugin Code
image

#### IDE Context Action
image

Apparently the IDE sorts them alphabetically...

**Comment by i582** [comment]
> Apparently the IDE sorts them alphabetically...

Yep, it's unclear if this can be fixed. In general, you can probably leave the current version, but maybe you can find a better solution.

**Comment by Danil42Russia** [comment]
> > Apparently the IDE sorts them alphabetically...
> 
> Yep, and it's unclear if this can be fixed. In general, you can probably leave the current version, but maybe you can find a better solution.

I can "support" with a crutch
image

**Comment by i582** [comment]
> > > Apparently the IDE sorts them alphabetically...
> > 
> > 
> > Yep, and it's unclear if this can be fixed. In general, you can probably leave the current version, but maybe you can find a better solution.
> 
> I can "support" with a crutch

Perhaps you can do it this way, because `array_first_value` is a more correct replacement for `reset`, and `array_first_key` is an additional replacement.

**Comment by Danil42Russia** [comment]
> > > > Apparently the IDE sorts them alphabetically...
> > > 
> > > 
> > > Yep, and it's unclear if this can be fixed. In general, you can probably leave the current version, but maybe you can find a better solution.
> > 
> > 
> > I can "support" with a crutch
> 
> Perhaps you can do it this way, because `array_first_value` is a more correct replacement for `reset`, and `array_first_key` is an additional replacement option.

That's what I did :)
image

**Comment by i582** [comment]
> That's what I did :)

By the way, it would be nice to add tests for quick fixes as well.

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        private val PHP_TO_KPHP_FUNCTIONS = mapOf(
+            "end" to listOf("array_last_value", "array_last_key"),
+            "reset" to listOf("array_first_value", "array_first_key"),
+        )
+    }
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : PhpElementVisitor() {
+            override fun visitPhpFunctionCall(reference: FunctionReference) {
+                if (reference.immediateNamespaceName != "")
```
Comment (i582) [inline src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt:23]: Why not just:

```kotlin
val functionName = reference.fqn
```

And:

```kotlin
private val PHP_TO_KPHP_FUNCTIONS = mapOf(
  "\\end" to listOf("array_last_value", "array_last_key"),
  "\\reset" to listOf("array_first_value", "array_first_key"),
)
```

This seems like a simpler solution and doesn't require any reference resolutions.

Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt:23]: Thanks to this code, we do not receive incorrect warnings:
```kotlin
if (reference.immediateNamespaceName != "" && reference.immediateNamespaceName != "\\")
    return

if ((reference as PhpReference).resolveLocal().isNotEmpty())
    return
```

```php
<?php # functions.php

namespace MyFunctions;

function reset(): string {
    return "sddsd";
}

$animal = array('dog', 'cat', 'fox');

echo reset(); // not warning 'KPHP does not support'
echo end($animal); // warning 'KPHP does not support'
echo \N\end($animal); // not warning 'KPHP does not support'
echo \reset($animal); // warning 'KPHP does not support'
```

```php
<?php # index.php

namespace N;

use MyFunctions;

function end(object|array &$array): string {
    return "";
}

$fruits = array('apple', 'banana', 'cranberry');

echo end($fruits); // not warning 'KPHP does not support'
echo MyFunctions\reset(); // not warning 'KPHP does not support'
echo reset($fruits); // warning 'KPHP does not support'
echo \reset($fruits); // warning 'KPHP does not support'
echo \end($fruits); // warning 'KPHP does not support'
```

```php
<?php # example.php

use function  N\end;
use function  N\end as reset;

$a = [];

echo end($a); // not warning 'KPHP does not support'
echo reset($a); // not warning 'KPHP does not support'
echo \reset($a); // warning 'KPHP does not support'
echo \end($a); // warning 'KPHP does not support'
```

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 6 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt]: Сopypasta ;) Fixed

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 5 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:29]: Fixed

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 4 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt]: Fixed

File: build.gradle (hunk 3 above)
Comment (Danil42Russia) [reply build.gradle]: Fixed

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 2 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt]: Fixed :)

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : PhpElementVisitor() {
+            override fun visitPhpFunctionCall(reference: FunctionReference) {
+                val functionName = reference.name
+                if (functionName !in PHP_TO_KPHP_FUNCTIONS.keys)
+                    return
+
+                val fixes = mutableListOf<LocalQuickFix>()
+
+                val maybeFunction = PHP_TO_KPHP_FUNCTIONS[functionName]
```
[the remark this answers is not in the document]
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt]: Fixed

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+                maybeFunction?.forEach { maybeFunctionName ->
```
[the remark this answers is not in the document]
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt]: It seems superfluous

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 1 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt]: After adding if it's not worth it :)

File: build.gradle (hunk 3 above)
Comment (Danil42Russia) [reply build.gradle]: Fixed

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 10]
```diff
@@ -8,33 +8,32 @@ import com.jetbrains.php.lang.psi.PhpPsiElementFactory
...
 
     override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
         val unsupportedFunction = descriptor.psiElement
-        if (unsupportedFunction !is FunctionReference) {
+        if (unsupportedFunction !is FunctionReference)
             return
-        }
 
-        val methodNameNode = unsupportedFunction.nameNode ?: return
+        val newFunctionNameNode = unsupportedFunction.nameNode ?: return
```
[the remark this answers is not in the document]
Comment (i582) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:27]: In its current form, this code seems confusing:

```kotlin
unsupportedFunction.node.replaceChild(
    newFunctionNameNode,
    newFunction
)
```

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt (hunk 7 above)
Comment (i582) [reply src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt]: Yep, it's true

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt (hunk 5 above)
Comment (i582) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:29]: It seems you mixed up lines 23 and 21, see the last comment.

**Comment by Danil42Russia** [comment]
@i582, can add an extension function to check for "global functions"?

```kotlin
fun FunctionReference.isGlobalFunction(): Boolean {
    if (this.immediateNamespaceName != "" && this.immediateNamespaceName != "\\")
        return false

    if ((this.reference as PhpReference).resolveLocal().isNotEmpty())
        return false

    return true
}
```

**Comment by i582** [comment]
> @i582, can add an **extension function** to check for "global functions"?
> 
> ```kotlin
> fun FunctionReference.isGlobalFunction(): Boolean {
>     if (this.immediateNamespaceName != "" && this.immediateNamespaceName != "\\")
>         return false
> 
>     if ((this.reference as PhpReference).resolveLocal().isNotEmpty())
>         return false
> 
>     return true
> }
> ```
> 
> Well, or in the **extension properties** :)

@Danil42Russia, I'm not sure if this is necessary, since you're not going to reuse this code somewhere else, I think it would be enough to move it to a private method of `object : PhpElementVisitor` to make the code more self-documenting.

**Comment by i582** [comment]
Found a funny bug:

Here is the error:
```php
<?php

$a = [];

echo reset($a);
```

But here it is no more:

```php
<?php

use function \reset;

$a = [];

echo reset($a);
```

Also code like this doesn't give errors either:

```php
<?php

use function \reset as reset2;

$a = [];

echo reset2($a);
```

And it looks a little strange:

Снимок экрана 2022-02-22 в 21 11 56

**Comment by Danil42Russia** [comment]
Fixed the bug that when replacing the function, the plugin did not delete the namespace

#### Before replacement
```php
<?php

$a = [];

echo \end($a); // warning 'KPHP does not support'
```

#### Before fixing the bug
```php
<?php

$a = [];

echo \array_last_value($a); // not warning
```

#### After fixing the bug
```php
<?php

$a = [];

echo array_last_value($a); // not warning
```

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (i582) [reply src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt]: Ok

File: src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElementVisitor
+import com.jetbrains.php.lang.inspections.PhpInspection
+import com.jetbrains.php.lang.psi.elements.FunctionReference
+import com.jetbrains.php.lang.psi.elements.PhpReference
+import com.jetbrains.php.lang.psi.visitors.PhpElementVisitor
+import com.vk.kphpstorm.inspections.quickfixes.ReplaceToKphpFunctionsQuickFix
+
+class KphpNotSupportFunctionsInspection : PhpInspection() {
```
Comment (unserialize) [inline src/main/kotlin/com/vk/kphpstorm/inspections/KphpNotSupportFunctionsInspection.kt:12]: `KphpNotSupportedFunctionsInspection`

File: src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.util.PsiTreeUtil
+import com.jetbrains.php.lang.lexer.PhpTokenTypes
+import com.jetbrains.php.lang.psi.PhpPsiElementFactory
+import com.jetbrains.php.lang.psi.elements.FunctionReference
+import com.jetbrains.php.lang.psi.elements.PhpNamespaceReference
+
+class ReplaceToKphpFunctionsQuickFix(
```
Comment (unserialize) [inline src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt:12]: ReplaceUnsupportedFunctionCallQuickFix

Comment (Danil42Russia) [reply src/main/kotlin/com/vk/kphpstorm/inspections/quickfixes/ReplaceToKphpFunctionsQuickFix.kt]: Fixed :)

**Review by unserialize — APPROVED** [review]

**Pull request merged**